### ISSUE TITLE
[browser] Module Loaders

### ISSUE THREAD

**Issue by kriskowal** [tlrobinson/narwhal#22]
Create a build system and module loaders for the browser.

Support:
- production: performance (minification, bundling, browser cache induction, parallelism)
- development: preserves line numbers and file names
- static: built with a tool and saved as static files
- dynamic: served by a JSGI application on demand, with an LRU Cache
- sync
- async

Methods:
- via async XHR: no global footprint, but restricted to same origin.
- via sync XHR: great for development on static files, but slow, serial, and buggy in production, and also lacks file names for debugging in browsers.
- via inline script; embeds a bundle so subsequent <scripts> may require modules.
- via blocking download script; downloads a bundle so subsequent <scripts> may require modules.
- via deferred script injection; downloads a bundle or individual modules asynchronously.  sync require not available for these modules in subsequent <scripts>, so it must throw programmer errors or use an alternate, dangerous sync XHR loader for those cases.

Consider:
- bundling speeds up initial page loads, but collections of smaller files cache better for faster following page loads
- browser cache induction uses mtime-based unique identifiers with far-future expiration
- file names and line numbers can be preserved by module factories that do not inject newlines at the head and foot of the module, if the factories are installed with script injection.
- if modules are loaded from multiple small files, those components can be loaded in parallel, as with labjs.  however, if modules are loaded asynchronously in non-deferred script tags, subsequent inline script tags will not be able to synchronously require those modules.
- after downloading a bundle, the client may initiate the download of the smaller factories for pre-caching; this only works if the modules are cacheable.  the server must track whether it expects the browser to have downloaded the smaller factories to decide whether to send the bundle loader or module loader in subsequent requests.

**Comment by kriskowal** [comment]
Also need to support the case where a module is expected to be loaded after blocking <script src> tag; only works if the script is a loader bundled with the module's transitive dependencies.
